Move IP de-allocation logic for Service into AfterDeleteFunc

MrHohn opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Ref https://github.com/kubernetes/kubernetes/issues/87603#issuecomment-579443444, the service registry currently doesn't handle object deletion in response to removing the last finalizer, specifically it fails to release the allocated IP.

This PR fixes that by moving the release logic into the AfterDelete hook.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #87603

**Special notes for your reviewer**:
/assign @liggitt @thockin 

I'm honestly not sure if I'm using the AfterDelete thing right, so feedbacks welcome :)

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fix a bug that service's clusterIP is not released promptly after deletion with finalizers.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

MrHohn commented:
/retest

MrHohn commented:
/retest

liggitt commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -261,8 +268,6 @@ func (rs *REST) Delete(ctx context.Context, id string, deleteValidation rest.Val
 			return nil, false, err
 		}
-
-		rs.releaseAllocatedResources(svc)
```
If I'm reading this correctly, the `endpoints.Delete` call should also be included in the AfterDelete handler, at which point I don't see a real reason for a custom REST#Delete implementation here.

The `rs.endpoints.Delete` and `rs.releaseAllocatedResources` calls are both done post-delete, which is already fragile if the server crashes between service delete and endpoints.Delete/releaseAllocatedResources calls, but I don't think moving these calls to AfterDelete makes that worse

liggitt replied:
Actually, this Delete handler should still have been called at the point at which the Delete API call was made to the service. Why weren't the resources released at that point, even if the service object in etcd wasn't deleted until the finalizer was removed?

liggitt commented:
This definitely needs tests exercising the case described in #87603:

* Create service+endpoints, verify IP is allocated and endpoints exist
* Update service to add custom finalizer
* Issue delete API call to service (should endpoints be deleted and resources released at this point or not?)
* Update service to remove custom finalizer
* Verify service no longer exists (should endpoints be deleted and resources released at this point or not?)

I'm unsure at what point in the service lifecycle the endpoints/resources are expected to be deleted/released:
* At the point when the Delete request is made and the deletionTimestamp is added to the service?
* At the point when the finalizer is removed and the service object is removed from etcd?

liggitt commented:
/sig network

needs input from sig-network about the expected service IP/resource lifecycle

MrHohn commented:
Humm, I was able to reproduce the IP not de-allocated issue with service deletion and re-creation with the same clusterIP even with my current change:
```
$ kubectl apply -f svc-nginx-loadbalancer.yaml 
service/my-nginx created
$ kubectl delete -f svc-nginx-loadbalancer.yaml 
service "my-nginx" deleted
$ kubectl apply -f svc-nginx-loadbalancer.yaml 
The Service "my-nginx" is invalid: spec.clusterIP: Invalid value: "10.0.203.27": provided IP is already allocated
```

Looking more into this.

MrHohn commented:
From the kube-apiserver logs, I see the sequence of:
- delete service from user
- patch service to remove the finalizer from service controller
- remove endpoint from service controller (why isn't this done by apiserver instead?)
- apiserver found out the clusterIP is leaked and cleaning up

Ref:
![image](https://user-images.githubusercontent.com/8681801/82105912-a728f800-96d2-11ea-89fe-b6aee11b76e2.png)

aojea commented:
> This definitely needs tests exercising the case described in #87603:
> 
> * Create service+endpoints, verify IP is allocated and endpoints exist
> * Update service to add custom finalizer
> * Issue delete API call to service (should endpoints be deleted and resources released at this point or not?)
> * Update service to remove custom finalizer
> * Verify service no longer exists (should endpoints be deleted and resources released at this point or not?)
> 
> I'm unsure at what point in the service lifecycle the endpoints/resources are expected to be deleted/released:
> 
> * At the point when the Delete request is made and the deletionTimestamp is added to the service?
> * At the point when the finalizer is removed and the service object is removed from etcd?

I have to admit that I tried to solve this riddle but I gave up due to the complexity (at least from my perspective) :sweat_smile: . Anyway, I created an integration test to try to reproduce it without success, but I think that it implements or can be adapted to fulfill Jordan requirements.

@liggitt do you think this test will cover it https://github.com/kubernetes/kubernetes/pull/91285 ?

If affirmative, @MrHohn can you incorporate it to this PR?

bowei commented:
/assign

thockin commented:
What's the status on this?  I have ignored it because I saw Bowei assign it, but is it on-track?  Does it need more attention?

bowei commented:
thanks for the ping :-) I think I need sync up with @MrHohn to understand this

MrHohn commented:
Sorry for my long long delay on this. I started working on this again, and after various local tests, I can totally see the releaseAllocatedResources() method was correctly triggered - as mentioned on https://github.com/kubernetes/kubernetes/pull/90753#discussion_r423231373 - in the right sequence during service deletion (with or without finalizer). I have no idea why repair controller still considers the IP and port were not allocated or leaked.
```
2020-07-16 21:59:24.803 PDT "[debug] create service default/my-nginx"
2020-07-16 21:59:24.828 PDT "[debug] Store.Update(name=my-nginx)"
2020-07-16 22:00:02.860 PDT "[debug] Store.Update(name=my-nginx)"
2020-07-16 22:00:31.619 PDT "[debug] deleting service my-nginx"
2020-07-16 22:00:31.621 PDT "[debug] deleted service default/my-nginx"
2020-07-16 22:00:31.621 PDT "[debug] deleting endpoint for service default/my-nginx"
2020-07-16 22:00:31.627 PDT "[debug] releasing IP for service default/my-nginx from Delete()"
2020-07-16 22:00:59.971 PDT "the cluster IP 10.0.203.27 for service my-nginx/default is not allocated; repairing"
2020-07-16 22:00:59.976 PDT "the node port 30547 for service my-nginx/default is not allocated; repairing"
2020-07-16 22:01:02.309 PDT "[debug] Store.Update(name=my-nginx)"
2020-07-16 22:01:02.313 PDT "kubelet config controller: deleteWithoutFinalizers - hit no error, calling finalizeDelete()"
2020-07-16 22:01:02.313 PDT "[debug] finalizeDelete(runHook=true) triggered for obj: &{TypeMeta:{Kind: APIVersion:} ObjectMeta:{Name:my-nginx GenerateName: Namespace:default SelfLink: UID:36596c6b-6fbf-4cef-980c-e8eb5c8cc995 ResourceVersion:1143 Generation:0 CreationTimestamp:2020-07-17 04:59:24 +0000 UTC DeletionTimestamp:2020-07-17 05:00:31 +0000 UTC DeletionGracePeriodSeconds:0xc00b3bcf98 Labels:map[run:my-nginx] Annotations:map[kubectl.kubernetes.io/last-applied-configuration:{"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"run":"my-nginx"},"name":"my-nginx","namespace":"default"},"spec":{"clusterIP":"10.0.203.27","ports":[{"nodePort":30547,"port":80,"protocol":"TCP"}],"selector":{"run":"my-nginx"},"type":"LoadBalancer"}} ] OwnerReferences:[] Finalizers:[service.kubernetes.io/load-balancer-cleanup] ClusterName: ManagedFields:[]} Spec:{Type:LoadBalancer Ports:[{Name: Protocol:TCP Port:80 TargetPort:{Type:0 IntVal:80 StrVal:} NodePort:30547}] Selector:map[run:my-nginx] ClusterIP:10.0.203.27 ExternalName: ExternalIPs:[] LoadBalancerIP: SessionAffinity:None SessionAffinityConfig:<nil> LoadBalancerSourceRanges:[] ExternalTrafficPolicy:Cluster HealthCheckNodePort:0 PublishNotReadyAddresses:false IPFamily:<nil> TopologyKeys:[]} Status:{LoadBalancer:{Ingress:[{IP:35.188.0.59 Hostname:}]}}}"
2020-07-16 22:01:02.318 PDT "[debug] Store.Update(name=my-nginx)"
2020-07-16 22:03:59.981 PDT "the cluster IP 10.0.203.27 may have leaked: flagging for later clean up"
2020-07-16 22:03:59.985 PDT "the node port 30547 may have leaked: flagging for later clean up"
2020-07-16 22:09:59.992 PDT "the cluster IP 10.0.203.27 appears to have leaked: cleaning up"
2020-07-16 22:09:59.995 PDT "the node port 30547 appears to have leaked: cleaning up"
```

That said, I don't think simply moving the de-allocation logic into AfterDeleteFunc will fix the issue and I will look further into this. Any help on investigation would be highly appreciated as well!

MrHohn commented:
Humm, looking at the timestamp closer - it looks like the repair controller kicked in between a service being deleted and its finalizers being removed. So the sequence might be:
- my-nginx service deleted with finalizer: clusterIP and port released
- repair controller checked on my-nginx service: mistakenly thought the clusterIP and port should be allocated and repaired them
- my-nginx service deleted for real after all finalizers detached: nothing happen here
- repair controller checked again: the previously repaired clusterIP and port were leaked.

MrHohn commented:
Okay, looks like fixing the repair controller to also check for deletionTimestamp before repairing resolves the IP/port leaking issue:
```
2020-07-17 11:12:36.179 PDT "[debug] deleting service my-nginx-cluster-ip"
2020-07-17 11:12:36.182 PDT "[debug] deleted service default/my-nginx-cluster-ip"
2020-07-17 11:12:36.182 PDT "[debug] deleting endpoint for service default/my-nginx-cluster-ip"
2020-07-17 11:12:36.195 PDT "[debug] releasing IP for service default/my-nginx-cluster-ip from Delete()"
2020-07-17 11:15:31.654 PDT "[new codes] svc default/my-nginx-cluster-ip is being deleted, skipping repairing IP..."
2020-07-17 11:15:31.657 PDT "[new codes] svc default/my-nginx-cluster-ip is being deleted, skipping repairing port..."
```

I will open a new PR and drop this one.

MrHohn commented:
Sent https://github.com/kubernetes/kubernetes/pull/93199 and closing this.

MrHohn committed:
Move resources de-allocation logic for Service into AfterDeleteFunc

MrHohn commented:
@liggitt I have updated the PR and moved the de-allocation logic for both endpoint and IP/port resources into the `AfterDeleteFunc` (per the feedback received from sig-network). PTAL thanks!

(Adding an E2E test in parallel but wanted to see if the implementation makes sense.)

MrHohn commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

aojea commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -242,18 +256,6 @@ func (rs *REST) Delete(ctx context.Context, id string, deleteValidation rest.Val
 
-	// Only perform the cleanup if this is a non-dryrun deletion
-	if !dryrun.IsDryRun(options.DryRun) {
-		// TODO: can leave dangling endpoints, and potentially return incorrect
-		// endpoints if a new service is created with the same name
```
does this no longer applies?

MrHohn replied:
Good catch, this is probably still true if we encounter unexpected error when deleting the endpoint. One thing that is not clear in my mind is whether or not the deletion of service will be completed if we hit such error in the afterDeleteHook, and whether or not a new service with the same name will be able to be created if the prior service deletion has not been completed. (The comments was added [a long while back](https://github.com/jennybuckley/kubernetes/commit/3ffe259ac773f5c9ec91d6e8b5751020a05af56b) and I suspect if it is stale.)

Nevertheless, this PR focuses more on moving the de-allocation step as is, so let me copy over the comments for now.

thockin commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -99,3 +104,8 @@ func (r *StatusREST) Update(ctx context.Context, name string, objInfo rest.Updat
 	return r.store.Update(ctx, name, objInfo, createValidation, updateValidation, false, options)
 }
+
+// SetAfterDeleteFunc sets the given after delete func.
```
Can these comments please explain why one sets r.AfterDelete and the other sets r.store.AfterDelete ?  This is very confusing to someone who isn't super familiar with the trickiness of Service

thockin commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -115,7 +126,10 @@ func NewREST(
 		pods:                pods,
 	}
-
+	// Configure the after delete func for both service and service status storage to ensure
```
This feels like a strange approach which isn't clearly documented.

You are defining a new type, then calling one method on it, passing the same argument to both types.  What purpose does the ServiceStatusStorage type serve?  Why not just set both AfterDelete functions in one method on one type?

I feel like there must be a reason, and I just don't see it.

MrHohn replied:
Agreed that this is strange. I have removed the unnecessary types and passing in an explicit `setAfterDelete` function so that it can set for both storages. PTAL if this is cleaner, thanks!

thockin commented on pkg/registry/core/service/storage/rest.go:
```diff
@@ -267,6 +269,22 @@ func (rs *REST) Delete(ctx context.Context, id string, deleteValidation rest.Val
+		}
+		rs.releaseAllocatedResources(svc)
+		ctx := genericapirequest.WithNamespace(context.Background(), svc.Namespace)
+		_, _, err := rs.endpoints.Delete(ctx, svc.Name, rest.ValidateAllObjectFunc, &metav1.DeleteOptions{})
```
Can you please comment this?  Why are we catching the AfterDelete hook, just to delete the object again?

aojea replied:
these delete the endpoints, we were deleting the service before, no?

aojea replied:
Actually, @thockin @liggitt , IIUIC,  it seems that this `endpoints.Delete()` should be removed because it will race with the endpoints controller

https://github.com/kubernetes/kubernetes/blob/4fbf5df52b122ef9f466676792b8afed7157ced5/pkg/controller/endpoint/endpoints_controller.go#L362-L367

another thing is that  this is inconsistent with endpointslices, that is the default since 1.19

liggitt replied:
> it seems that this `endpoints.Delete()` should be removed because it will race with the endpoints controller

Even though I agree it is redundant in theory, I'm hesitant to remove the synchronous delete that has been in place since 1.0.

> another thing is that this is inconsistent with endpointslices, that is the default since 1.19

EndpointSlices is still in beta, I would not change the GA behavior to match the beta behavior

thockin reviewed (commented):
I'm trying to understand this all (after some time away from this code) and failing.  It desperately needs comments.

MrHohn reviewed (commented):
Thanks for the reviews! I have updated the codes, PTAL.

MrHohn commented on pkg/registry/core/service/storage/storage.go:
```diff
@@ -56,12 +58,19 @@ func NewGenericREST(optsGetter generic.RESTOptionsGetter, serviceCIDR net.IPNet,
-	return &GenericREST{store}, &StatusREST{store: &statusStore}, nil
+	// Returning a function such that the caller will be able to customize the AfterDelete
+	// step for the service resource as necessary. AfterDelete should implement a further
+	// operation to run after a service is deleted and before it is decorated.
```
I can't find specific public document around this AfterDelete operation. The closest is the comment from the registry.Store field. So I put something similar here.
https://github.com/kubernetes/kubernetes/blob/09cd56b561d0c31421dfc6374b68e0b18c6cc746/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L163-L165

MrHohn committed:
Re-generate openapi spec for service/status

MrHohn commented:
Looks like after editing StatusREST some tests broke:
```
--- FAIL: TestStorageVersionHashes (0.21s)
    instance_test.go:400: expect resource v1/services/status to have empty storageVersionHash, got hash "0/CO1lhkEBI="
```

Need to investigate more.

khenidak commented:
@MrHohn @thockin 

While I really appreciate this. If this PR merged my life on dualstack will become substantially more complex. We have modified rest/storage so much for dual stack.. please check https://github.com/kubernetes/kubernetes/pull/95239 and let us talk about a strategy to align the two PRs together

MrHohn commented:
@khenidak Thanks for chiming in. Totally agreed that we should see how to align this with the dualstack effort. I took a quick glance on https://github.com/kubernetes/kubernetes/pull/95239 and I don't see any changes on rest/storage. Did you actually mean https://github.com/kubernetes/kubernetes/pull/91824?

khenidak commented:
> #91824

Yes. That is the correct PR. The commit of interest is the alloc and repair ones.

MrHohn commented:
> Yes. That is the correct PR. The commit of interest is the alloc and repair ones.

Thanks for confirming. Took a first pass on it. In fact, I don't see too much of major conflict out of the changes. #91824 seems to change a lot of the cluster IP allocation & de-allocation logic related to the IP families. Some codes are added to service storage and strategy for defaulting and normalizing the service object.

This PR mostly changes the sequence of the de-allocation step without caring about how exactly the de-allocation will happen. So my understanding is this is orthogonal (feel free to point out anything I missed!).

If the concern is conflict of lines - I'm totally fine with holding and rebasing after #91824 is merged.

cmluciano commented:
@MrHohn looks like the dual stack services PR merged, do you need any help with progressing on this PR?

MrHohn commented:
> @MrHohn looks like the dual stack services PR merged, do you need any help with progressing on this PR?

Thanks for the update, great and I will rebase this PR.

khenidak commented:
@MrHohn + @thockin 

Tim is working on moving thing to strategy with a clearer cleanup approach -- These two PRs will conflict with each others. I suggest a huddle and agreeing on a unified approach

PR: https://github.com/kubernetes/kubernetes/pull/95967#issuecomment-724174624
